### PR TITLE
Resolves #704: Allow user to set custom information in the store header

### LINKED ISSUE #704: Allow user to set custom information in the store header
The store header is read every time the user opens a record store, and there are already mechanisms for caching it in ways that are transactionally consistent with FDB. As a result, it would be convenient if the user could specify a few custom fields on the record, though there should be appropriate warnings that this should be used for read-heavy data with very low write rates.

I think the approach I want this to be is a `map<string, bytes>`. Given that in proto2, maps don't exist, we'll probably need to fake it with a nested protobuf message with the same semantics as a map. Then we can fake it by setting the field(s) ourselves. Then the user can serialize whatever data they want into the value (in their favorite serialization format), and you can even have different users with different types of data sitting next to each other.

In theory, this should be, like, `map<string, Any>` or even just an `Any`. This is more "structured", and it allows for a more formal mechanism of type checking, etc. But `Any` is only available in proto3, and faking it in proto2 is...more complicated. So I'll probably stick to `bytes` for now.

### PR DESCRIPTION
**Pull request by alecgrieser**
This allows the user to set fields in the store header. The exact implementation is something that is wire (and semantics) compatible with a `map<string, bytes>`, but not a `map<string, bytes>` because that feature is not available in proto2. I think the way people should be encouraged to use this is to use protobuf as the serialization mechanism, and then they can use standard proto evolution guidelines to evolve things. In theory, a single user might have separate "top level keys" that they want to set, so they can, but generally, they should keep this relatively small. I've included some caveats and some usage examples in the documentation.

### CODE REVIEW

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/FDBRecordStore.java [hunk 1]
```diff
@@ -2089,6 +2093,210 @@ private boolean isStateCacheableInternal() {
...
+     *
+     * @param userField the name of the user-settable field to read
+     * @return the value of that field in the header or {@code null} if it is unset
+     * @see #setHeaderUserFieldAsync(String, ByteString)
+     */
+    @Nullable
+    public ByteString getHeaderUserField(@Nonnull String userField) {
+        validateCanAccessHeaderUserFields();
+        if (recordStoreStateRef.get() == null) {
+            throw new RecordCoreException("cannot get field from header as record store state is null");
```
Comment (scgray) [inline fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/FDBRecordStore.java]: For this sort of exception, it there any guidance we could give in the text as to what this means or what could be done with it?  Or, at least, if the store state being `null` is a unexpected state to be found in?   Also, maybe this should include logging info of the subspace provider of the store.

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/FDBRecordStore.java [hunk 2]
```diff
@@ -2089,6 +2093,210 @@ private boolean isStateCacheableInternal() {
...
+     * </p>
+     *
+     * @param userField the name of the user-settable field to write
+     * @param value the value to set the field to
+     * @return a future that will be ready when setting the field has completed
+     * @see #getHeaderUserField(String)
+     */
+    @Nonnull
+    public CompletableFuture<Void> setHeaderUserFieldAsync(@Nonnull String userField, @Nonnull ByteString value) {
+        return updateStoreHeaderAsync(storeHeaderBuilder -> {
```
Comment (scgray) [inline fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/FDBRecordStore.java]: The way this is written, if you set a user field it won't be visible from `getHeaderUserField()` until the transaction has been committed.  Perhaps this should have a javadoc to this effect, or this method should actually update the in-memory version prior to returning the future?   I think the former probably makes more sense.

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/FDBRecordStore.java [hunk 3]
```diff
@@ -2089,6 +2093,210 @@ private boolean isStateCacheableInternal() {
...
+     * @see #setHeaderUserFieldAsync(String, ByteString)
+     * @see #setStateCacheabilityAsync(boolean)
+     */
+    public void setHeaderUserField(@Nonnull String userField, @Nonnull byte[] value) {
+        context.asyncToSync(FDBStoreTimer.Waits.WAIT_EDIT_HEADER_USER_FIELD, setHeaderUserFieldAsync(userField, value));
+    }
+
+    /**
+     * Clear the value of a user-settable field in the store header. This removes a field from the store header
+     * after it has been set by {@link #setHeaderUserFieldAsync(String, ByteString)}. This has the save caveats
```
Comment (scgray) [inline fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/FDBRecordStore.java]: `save` -> `same`

**Comment by alecgrieser** [comment]
Talked to @shralex offline, and he was okay with these changes.

**Pull request merged**